fetch/mesh_data.go

```diff
@@ -56,19 +56,40 @@ func (f *Fetch) getHashes(ctx context.Context, hashes []types.Hash32, hint datas
 				continue
 			}
 			if !res.IsLocal {
-				if err := receiver(ctx, res.Data); err != nil {
+				err := receiver(ctx, res.Data)
+				if err != nil {
 					f.logger.WithContext(ctx).With().Warning("failed to handle data",
 						log.String("hint", string(hint)),
 						log.String("hash", hash.String()),
 						log.Err(err))
 					errs = append(errs, err)
 				}
+				f.doneProcessing(res, err)
 			}
 		}
 	}
 	return errs
 }
 
+func (f *Fetch) doneProcessing(res HashDataPromiseResult, err error) {
+	f.activeReqM.Lock()
+	defer f.activeReqM.Unlock()
+	reqs := f.pendingRequests[res.Hash]
+	// doneProcessing is called when the first request of the hash finished validating.
+	// hand the data to more requests if any
+	if len(reqs) > 1 {
+		for _, req := range reqs[1:] {
```

Comment:
i think this is very hackish and prone to misuse, there should be a better approach

Reply:
i would try to make it work with only one request instead of collection so that it would be definitely correct

```
type Promise struct {
     completed chan struct{}
     value Result
}
```

single writer updates value after validating it and closes completed channel
all concurrent readers are waiting on the promise

Reply:
changed to what you suggested